File: src/main/java/de/istkorrekt/gameoflife/Location.java
package de.istkorrekt.gameoflife;

import java.util.Comparator;
import java.util.Objects;

public final class Location implements Comparable<Location> {

    private final int x;
    private final int y;

    private Location(int x, int y) {
        this.x = x;
        this.y = y;
    }

    public int getX() {
        return x;
    }

    public int getY() {
        return y;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Location location = (Location) o;
        return x == location.x &&
                y == location.y;
    }

    @Override
    public int hashCode() {
        return Objects.hash(x, y);
    }

    @Override
    public int compareTo(Location location) {
        return Comparator.comparing(Location::getX).thenComparing(Location::getY).compare(this, location);
    }

    @Override
    public String toString() {
        return "(" + x + "|" + y + ")";
    }

    public boolean isWithinBounds(Size size) {
        // x and y are zero-based
        return x < size.getWidth() && y < size.getHeight();
    }

    public Location upper() {
        int upperY = Math.max(y - 1, 0);
        return Location.of(x, upperY);
    }

    public Location right(Size bounds) {
        Location newLocation = Location.of(x + 1, y);
        if (newLocation.isWithinBounds(bounds)) {
            return newLocation;
        }
        return this;
    }

    public Location lower(Size bounds) {
        Location newLocation = Location.of(x, y + 1);
        if (newLocation.isWithinBounds(bounds)) {
            return newLocation;
        }
        return this;
    }

    public Location left() {
        int update = Math.max(x - 1, 0);
        return Location.of(update, y);
    }

    public static Location of(int x, int y) {
        return new Location(x, y);
    }
}

File: src/main/kotlin/de/istkorrekt/gameoflife/Location.kt
package de.istkorrekt.gameoflife

data class Location(val x: Int, val y: Int) : Comparable<Location> {

    companion object {
        private val COMPARATOR = Comparator.comparingInt<Location> { it.x }
                .thenComparingInt { it.y }
    }

    override fun compareTo(other: Location): Int {
        return COMPARATOR.compare(this, other)
    }

    override fun toString(): String {
        return "($x|$y)"
    }

    fun upper(): Location? {
        val upperY = (y - 1).coerceAtLeast(0)
        return Location(x, upperY)
    }

    fun right(bounds: Size): Location {
        // x and y are zero-based
        val rightX = (x + 1).coerceAtMost(bounds.width - 1)
        return Location(rightX, y)
    }

    fun lower(bounds: Size): Location {
        // x and y are zero-based
        val lowerY = (y + 1).coerceAtMost(bounds.height - 1)
        return Location(x, lowerY)
    }

    fun left(): Location {
        val update = (x - 1).coerceAtLeast(0)
        return Location(update, y)
    }
}